### PR TITLE
Jc/per project deps in properties

### PR DESCRIPTION
**Pull request by jmcampanini**
// for feedback: @uschi2000 + @qinfchen

it adds project specific syntax to the properties file

#### TODO
- [x] maybe improve the syntax for the project specific entry? it's `-projectName-` right now :(
- [x] add README - nothing in the README, would need to modify the wiki :(
- [ ] check if it works in their CI - for some reason build fails locally, but tests pass in IDE
- [ ] write an integration test for a multi-project setup

### CODE REVIEW

File: src/test/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProviderSpec.groovy [hunk 1]
```diff
@@ -43,12 +63,53 @@ class PropertyFileRecommendationProviderSpec extends Specification {
...
+        project = ProjectBuilder.builder().withName('projectName').build()
+        provider = new PropertyFileRecommendationProvider(project)
+
+        propFile << '''
+            GUAVA_VERSION = 18.0
+            com.google.guava:guava = $GUAVA_VERSION
+            some:other = $com.google.guava:guava
+            -notProjectName-GUAVA_VERSION = 11.0
+            -projectName-GUAVA_VERSION = 99.0
+            -projectName-com.google.guava:guava = $GUAVA_VERSION
```
Comment (jmcampanini) [inline src/test/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProviderSpec.groovy]: it actually allows for per-project variables to be resolved.. nifty!

File: src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java [hunk 2]
```diff
@@ -12,14 +12,20 @@
...
         @Override
         protected Collection<String> propertyNames() {
             return recommendations.stringPropertyNames();
         }
 
         @Override
         protected String propertyValue(String name) {
-            return recommendations.getProperty(name);
+            String key = "-" + project.getName() + "-" + name;
+            String rv = recommendations.getProperty(key);
```
Comment (uschi2000) [inline src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java]:
````
return Objects.firstNonNull(recommendations.getProperty(key), recommendations.getProperty(name));
````

File: src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java [hunk 3]
```diff
@@ -12,14 +12,20 @@
...
+
         @Override
         protected Collection<String> propertyNames() {
             return recommendations.stringPropertyNames();
         }
 
         @Override
         protected String propertyValue(String name) {
-            return recommendations.getProperty(name);
+            String key = "-" + project.getName() + "-" + name;
```
Comment (uschi2000) [inline src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java]: so the plugin already had support for the `-project-group:module` syntax, it just wasn't, uhm, "enabled"?

File: src/test/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProviderSpec.groovy [hunk 4]
```diff
@@ -17,18 +17,38 @@ class PropertyFileRecommendationProviderSpec extends Specification {
...
     }
 
+    def 'per-project exact match recommendations are provided'() {
+        when:
+        project = ProjectBuilder.builder().withName('projectName').build()
+        provider = new PropertyFileRecommendationProvider(project)
+
+        propFile << '''
+            com.google.guava:guava = 18.0
+            -notProjectName-com.google.guava:guava = 99.0
```
Comment (uschi2000) [inline src/test/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProviderSpec.groovy]: should be != 99

File: src/test/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProviderSpec.groovy [hunk 5]
```diff
@@ -17,18 +17,38 @@ class PropertyFileRecommendationProviderSpec extends Specification {
...
-        provider = new PropertyFileRecommendationProvider(null)
+        provider = new PropertyFileRecommendationProvider(project)
     }
 
     def 'exact match recommendations are provided'() {
         when:
-        propFile << 'com.google.guava:guava = 18.0'
+        propFile << '''
+            com.google.guava:guava = 18.0
+            -notProjectName-com.google.guava:guava = 99.0
```
Comment (uschi2000) [inline src/test/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProviderSpec.groovy]: should be != 99

**Comment by uschi2000** [comment]
this is cool, let's propose it to them.

File: src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java (hunk 3 above)
Comment (jmcampanini) [inline src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java]: yea. they had all the pieces, just didn't put it together that way...

File: src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java (hunk 2 above)
Comment (jmcampanini) [inline src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java]: the method can and should return `null` for it to function properly, and firstNoNull throws an NPE if both are null :(

Comment (uschi2000) [inline src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java]: oh, I see

**Comment by jmcampanini** [comment]
addressed your comments. going to modify the README and write an integration test for a multi-project setup, just to make sure it works. then will submit a PR to nebula

**Comment by uschi2000** [comment]
nice

File: src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java [hunk 6]
```diff
@@ -19,7 +21,12 @@
 
         @Override
         protected String propertyValue(String name) {
-            return recommendations.getProperty(name);
+            String projectSpecificName = project.getName() + PROJECT_DELIM + name;
```
Comment (qinfchen) [inline src/main/groovy/netflix/nebula/dependency/recommender/provider/PropertyFileRecommendationProvider.java]: what does this mean for root project? how does this play with global settings? is there any difference?

**Pull request closed without merging**